### ISSUE TITLE
Using TryAsync in code that uses async/awayt

### ISSUE THREAD

**Issue by SeanFarrow** [louthy/language-ext#255]
Hi,

I've currently got asp.net mvc core code that looks like the following:
```c#
    public async Task<IActionResult> GetGendersAsync()
        {
            try
            {
                var genders = await _repository.GetAllGendersAsync();
                return Ok(genders);
            }
            catch (Exception ex)
            {
                return StatusCode((int)HttpStatusCode.InternalServerError);
            }
                }
```
Given that my controller action is async and uses an await, I'm not sure how I should be converting this to use TryAsync, or if it is even possible, so would ask what the best approach is?

Secondly, in the TryAsync constructor function you are doing:
        public static TryAsync<A> TryAsync<A>(Func<A> f) => 
            TryAsyncExtensions.Memo<A>(() =>
                Task.Run(() => new Result<A>(f())));

Given .net 4.5.1 is the target framework, is there any reason Task.FromResult is not being used as Task.Run doesn't work well with asp.net web api according to this blog post:
[Async Programming - Introduction to Async/Await on ASP.NET](https://msdn.microsoft.com/en-us/magazine/dn802603.aspx)

I'm happy to make the change if it's something that would be warranted.
Kind regards
Sean.

**Comment by trbngr** [comment]
Sean,

This should work for you:

```csharp
        public async Task<IActionResult> GetGendersAsync() =>
            Try(await _repository.GetAllGendersAsync()).Match(
                Succ: Ok,
                Fail: e => (IActionResult) StatusCode(500)
            );
```

**Comment by SeanFarrow** [comment]
Hi Chris,

That worked a treat, so much so that I'm going to pass you some more code that is slightly more complicated.

I'm currently returning an Option<Gender from my repo, would returning a TryOption<Gender> help as I throw exceptions, or does it not really make a difference?
I'm now looking to port the following code as suctsinkly as possible:
public async Task<IActionResult> GetGenderByIDAsync(Guid id)
{
if (id == default(Guid))
            {
                return (IActionResult)BadRequest();
            }

            try
            {
                var gender =ifNone(await _repository.GetGenderByIDAsync(id), () =>new Gender { Id =default(Guid), Name =String.Empty});

                if (gender.Id == default(Guid) && !gender.Name.Any())
                {
                    //No gender with the passed in Id was found.
                    return NotFound();
                }

                return Ok("test");
            }
            catch (Exception ex)
            {
                return StatusCode((int)HttpStatusCode.InternalServerError);
            }
        }

Any help appreciated as I'm new to the library, but can really see it's uses.
Cheers
Sean.

**Comment by trbngr** [comment]
Hey Sean,

I decided to dive on in ;) Here's a repo for you

https://github.com/trbngr/langext-issue255/blob/master/langext-issue255/UnitTest1.cs

@louthy can you explain why the `monadic` approach [here](https://github.com/trbngr/langext-issue255/blob/master/langext-issue255/UnitTest1.cs#L84) isn't working? It's throwing `TaskCancelledExceptions`  for the tests `no_gender_found_returns_not_found` and `default_guid_returns_not_found`.

Cheers,
Chris

**Comment by louthy** [comment]
> can you explain why the monadic approach here isn't working?

Seems there was an issue in the bind for `MTryOptionAsync`, should be fixed in `v2.1.23`
